folium/colormap.py

```diff
@@ -0,0 +1,594 @@
+        if x[0]=="#" and len(x)==7:
+            # Color of the for #RRGGBB
+            color_tuple = (int(x[1:3],16), int(x[3:5],16), int(x[5:7],16))
+        else:
+            color_code = _cnames.get(x.lower(),None)
+            if color_code is None:
+                raise ValueError('Unknown color.')
+            color_tuple = (int(color_code[1:3],16), int(color_code[3:5],16), int(color_code[5:7],16))
+    if max(color_tuple)>1.:
+        color_tuple = tuple(u/255. for u in color_tuple)
+    return tuple(map(float,(color_tuple+(1.,))[:4]))
+
+_round = round
+
+def _base(x):
+    if x > 0:
+        base = pow(10, math.floor(math.log10(x)))
+        return round(x/base)*base
+    else:
+        return 0
+
+class ColorMap(MacroElement):
+    """A generic class for creating colormaps."""
+
+    def __init__(self, min=0., max=1., caption=""):
+        """
+        """
+        super(ColorMap, self).__init__()
+        self._name = 'ColorMap'
+
+        self.min = min
+        self.max = max
+        self.caption = caption
+        self.index = [min, max]
+
+        self._template = self._env.get_template('color_scale.js')
+
+    def render(self, **kwargs):
+        self.color_domain = [self.min + (self.max-self.min)*i/499. for i in range(500)]
+        self.color_range = [self.__call__(x) for x in self.color_domain]
+        self.tick_labels = legend_scaler(self.index)
+        self.caption = ""
+
+        super(ColorMap, self).render(**kwargs)
+
+        figure = self.get_root()
+        assert isinstance(figure, Figure), ("You cannot render this Element "
+                                            "if it's not in a Figure.")
+
+        figure.header.add_children(
+            JavascriptLink("https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min.js"),  # noqa
+            name='d3')
+
+    def _rgba_floats_tuple(self, x):
+        """This class has to be implemented for each class inheriting from Colormap.
+        This has to be a function of the form float -> (float,float,float,float)
+        Describing for each input float x, the output color in RGBA format ;
+        each output value being between 0 and 1.
+        """
+        raise NotImplementedError
+
+    def rgba_floats_tuple(self, x):
+        """Provides the color corresponding to value `x` in the
+        form of a tuple (R,G,B,A) with float values between 0. and 1.
+        """
+        return self._rgba_floats_tuple(x)
+
+    def rgba_bytes_tuple(self, x):
+        """Provides the color corresponding to value `x` in the
+        form of a tuple (R,G,B,A) with int values between 0 and 255.
+        """
+        return tuple(int(u*255.9999) for u in self.rgba_floats_tuple(x))
+
+    def rgb_bytes_tuple(self, x):
+        """Provides the color corresponding to value `x` in the
+        form of a tuple (R,G,B) with int values between 0 and 255.
+        """
+        return self.rgba_bytes_tuple(x)[:3]
+
+    def rgb_hex_str(self, x):
```

Comment:
When I see all these methods I cannot stop thinking about a tiny lib for colormaps... I've seen implementations like this in so many places now...